### ISSUE TITLE
Potential Memory leak?

### ISSUE THREAD

**Issue by TheWithz** [wavebox/waveboxapp#453]
* Which version of Wavebox are you using? [e.g. 3.0.0]
3.5.1 beta
* Which Operating System are you using? [e.g. Windows 10, Mac OSX 10, Ubuntu 14.04]
[Solus 3](https://solus-project.com)

So, I checked my email this morning on Wavebox, like I do every morning, then went to my classes and just returned. `so about 9 hours.` I go on my computer and see this

>thewithz:@:thewithz_solus:[/home/.../lib/Wavebox-linux-x64]:[14:03:38]> ./Wavebox
>[WB_LAS] started 30627
>[WB_LAS] no other processes found. Allowing run...
>[31100:1102/154316.606469:ERROR:stunport.cc(88)] Binding request timed out from 130.49.52.x:53971 (eno1)
>[30627:1102/161705.802085:FATAL:memory_linux.cc(35)] Out of memory.
>#0 0x000000b484b7 <unknown>
>#1 0x000000b5cb8b <unknown>
>#2 0x000000b27d1e <unknown>
>#3 0x000000abaeb4 operator new()
>#4 0x000001f18342 <unknown>
>#5 0x000001f1367e <unknown>
>#6 0x000001f0adfb <unknown>
>#7 0x000000acd626 <unknown>
>#8 0x000000b3af66 <unknown>
>#9 0x000000b6145d <unknown>
>#10 0x000000b6200e <unknown>
>#11 0x000000b4600d <unknown>
>#12 0x000000b61177 <unknown>
>#13 0x000000b0f87e <unknown>
>#14 0x000001f83245 <unknown>
>#15 0x000001f833f3 <unknown>
>#16 0x000000b15437 <unknown>
>#17 0x000000b4e513 <unknown>
>#18 0x7f9461d317d7 <unknown>
>#19 0x7f945dd910ef clone

>Aborted

I have 64 gigabytes of ram...

**Comment by Thomas101** [comment]
Wow, that shouldn't be happening. A few things will probably help us to diagnose...

1. Which install method did you use to install? (e.g. deb, snap, tar etc...)
2. Which accounts have you got added into Wavebox and what's the sleep status of each? (sleep can be found under settings for each account and service) e.g....
  * Gmail (no sleep)
    * Calendar (no sleep)
    * Notes (sleep)
  * Web link (google.com)(no sleep)
  * ...
3. Do you have any extensions enabled, if so which ones?

Also, if you launch Wavebox and open up the settings there's a task monitor button. This will display which tabs are using what resources. If you run Wavebox for a while or revisit, do any of these start to creep up?

**Comment by TheWithz** [comment]
Hello, sorry for the late follow-up. I was away for the whole weekend and did not have access to my computer. 
>Method: unzipped the latest beta tar.gz (3.5.1)

>Accounts:

1. Inbox (no sleep)
    - Calendar (no sleep)
    - Drive
    - Drive
    - Drive
    - Drive
    - Drive
    - Drive
    - Drive
2. Four Slack accounts
3. Two weblinks

>Extensions:

1. markdownHere
2. Gmelius

**Comment by TheWithz** [comment]
Hello, sorry for the late follow-up. I was away for the whole weekend and did not have access to my computer. 
>Method: unzipped the latest beta tar.gz (3.5.1)

>Accounts:

1. Inbox (no sleep)
    - Calendar (no sleep)
    - Drive (sleep 300)
    - Keep (sleep 300)
    - Photos (sleep 300)
    - Hangouts (no sleep)
    - Contacts (sleep 300)
    - Docs (sleep 300)
    - Sheets (sleep 300)
    - Slides (sleep 300)
    - Analytics (sleep 300)
    - Youtube (sleep 300)
    - Allo (sleep 300)
2. Inbox (no sleep)
    - Calendar (no sleep)
    - Drive (sleep 300)
    - Keep (sleep 300)
    - Photos (sleep 300)
    - Hangouts (no sleep)
    - Contacts (sleep 300)
    - Docs (sleep 300)
    - Sheets (sleep 300)
    - Slides (sleep 300)
    - Analytics (sleep 300)
3. Slack (no sleep)
4. Slack (no sleep)
5. Slack (no sleep)
6. Slack (no sleep)
7. Weblink (thewithz.tech)(sleep 300)
8. Weblink (reddit.com)(sleep 300)

>Extensions:

1. markdownHere
2. Gmelius

**Comment by TheWithz** [comment]
As far as the Task Monitor goes, I've only seen it after I come back from classes after being gone all day, and by then the computer has either restarted or the app has crashed, which is what gave the error output in the terminal.

**Comment by Thomas101** [comment]
Thanks for the info! I wonder if it's a memory leak in some of the hosted content. Would you be able to take a snapshot of the task monitor and then say 6 or 12 hours later do that again? We can have a look to see if anything has creeped up then

**Comment by TheWithz** [comment]
This is the task list now, I'll leave it on while I go to classes today. 
![screenshot from 2017-11-07 08-57-33](https://user-images.githubusercontent.com/8269545/32497273-05ba7042-c39a-11e7-86fe-8e1455079ca5.png)
![screenshot from 2017-11-07 08-57-43](https://user-images.githubusercontent.com/8269545/32497274-05c5c76c-c39a-11e7-9eba-3bdeefd16025.png)

**Comment by TheWithz** [comment]
just arrived back from classes to find that Wavebox has crashed again, which means I cannot see the task monitor. This is the error from the terminal, same as the previous one. 

>[WB_LAS] started 29390
>[WB_LAS] no other processes found. Allowing run...
>[29774:1106/172807.255332:ERROR:stunport.cc(88)] Binding request timed out from [2620:102:401f`:x:x:x:x:x`]:46773 (eno1)
>[29774:1107/143003.002643:ERROR:stunport.cc(88)] Binding request timed out from 130.49.52.x:43629 (eno1)
>[29390:1107/145004.663334:FATAL:memory_linux.cc(35)] Out of memory.
>#0 0x000000b484b7 <unknown>
>#1 0x000000b5cb8b <unknown>
>#2 0x000000b27d1e <unknown>
>#3 0x000000abaeb4 operator new()
>#4 0x000001f18342 <unknown>
>#5 0x000001f1367e <unknown>
>#6 0x000001f0adfb <unknown>
>#7 0x000000acd626 <unknown>
>#8 0x000000b3af66 <unknown>
>#9 0x000000b6145d <unknown>
>#10 0x000000b6200e <unknown>
>#11 0x000000b4600d <unknown>
>#12 0x000000b61177 <unknown>
>#13 0x000000b0f87e <unknown>
>#14 0x000001f83245 <unknown>
>#15 0x000001f833f3 <unknown>
>#16 0x000000b15437 <unknown>
>#17 0x000000b4e513 <unknown>
>#18 0x7fad7ee977d7 <unknown>
>#19 0x7fad7aef70ef clone

>Aborted

**Comment by TheWithz** [comment]
is there a way I can modify the Wavebox build with a simple if statement. Something like `if(memory used by [Pid] is > 1GB) print [name] of [Pid]` I chose 1GB arbitrarily but since something is eating in excess of 64 GB... 1 is small potatoes.

**Comment by Thomas101** [comment]
Open Dev tools with <kbd>Ctrl</kbd>+<kbd>Shift</kbd>+<kbd>J</kbd>, click on the console tab and paste the following...

```js
setInterval(() => {
  const electron = require('electron')
  const fs = require('fs')
  const metrics = electron.remote.app.getAppMetrics()
  const allWc = electron.remote.webContents.getAllWebContents()
  const indAllWc = allWc.reduce((acc, wc) => {
    acc.set(wc.getOSProcessId(), wc)
    return acc
  }, new Map())


  const big = metrics.find((m) => {
    return m.memory.workingSetSize > 1024*1000
  })

  if (big) {
    const out = metrics.map((m) => {
	  return Object.assign(m, {
         url: indAllWc.has(m.pid) ? indAllWc.get(m.pid).getURL() : '?'
      })
    })
    fs.writeFileSync('/Users/tom/Desktop/big', JSON.stringify(out,null,2))
	electron.remote.app.quit()
  }
}, 5000)
```

Make sure you change the path it's writing out to. You'll get a nice big dump of all the processes and the urls for them where applicable. The app will also quit when the process is big

**Comment by TheWithz** [comment]
I left Wavebox on overnight and was about to post the new function into the console when I saw this. Not sure if you can glean any more new info from it. 
![screenshot from 2017-11-08 09-07-12](https://user-images.githubusercontent.com/8269545/32553375-bd728ede-c464-11e7-82ef-254c924a3c21.png)

**Comment by Thomas101** [comment]
If you've got more than 10 tabs open at any one time that's normal. Nobodies got around to increasing the max listener number yet

**Comment by TheWithz** [comment]
Hmm... came back from classes today to find my computer shut off, Got excited because the log should have been able to catch what the culprit was. Boot up, look at the file... It's empty. Not sure what went wrong. I changed the path to correct file. `/home/thewithz/big.txt`

**Comment by Thomas101** [comment]
I guess the other thing to try is just writing the log every x minutes. We can assume the last one is when it crashed. This should just write out every minute. If there's any process grown out of control we should be able to see it!

```js
setInterval(() => {
  const electron = require('electron')
  const fs = require('fs')
  const metrics = electron.remote.app.getAppMetrics()
  const allWc = electron.remote.webContents.getAllWebContents()
  const indAllWc = allWc.reduce((acc, wc) => {
    acc.set(wc.getOSProcessId(), wc)
    return acc
  }, new Map())


  const out = metrics.map((m) => {
	return Object.assign(m, {
       url: indAllWc.has(m.pid) ? indAllWc.get(m.pid).getURL() : '?'
    })
  })
  fs.writeFileSync('/Users/tom/Desktop/big', JSON.stringify(out,null,2))
}, 60000)
```

**Comment by TheWithz** [comment]
Just started the log, It is populating the text file correctly. Now to wait!

**Comment by TheWithz** [comment]
Came back from classes to find Wavebox has crashed again, here is the `big.txt`
[big.txt](https://github.com/wavebox/waveboxapp/files/1459434/big.txt)
 file

**Comment by Thomas101** [comment]
Thanks for the log! Interestingly it looks like none of your accounts are sleeping. There's a process setup and running for each. The total memory consumption is at 8gigs which is far far away from your 64gig limit.

As regards the sleeping accounts, can you give this a try....
* Reduce the sleep timeout of one of the services (e.g. the first drive account) to be something low like 10
* Restart Wavebox
* Open the Task Monitor from settings - you should see 14ish processes from the config you posted above.
* Open the service we changed. You should see it appear in the processes
* Open a service that's already awake
* After 10 seconds the service that's set to sleep should disappear

Another thing that would be good to try, is before you're ready to leave your machine...
* Restart Wavebox
* Confirm there's 14-ish processes running again
* Add the logging code again
* Leave it in this state. Does that still crash and what's the log from that.

Another thought we've had in the office, is that there may be an upper memory limit on electron. Say 8gigs. In normal use that's not a problem because there would be memory pressure before the limit is reached, but obviously with 64 gigs this isn't going to happen. I'll see if I can trigger a low memory warning artificially and we can try that in the next beta see if it has any effect.

Thanks

**Comment by TheWithz** [comment]
I awakened them all because I wanted to see if any applications that were previously sleeping had the same problems. I would not have done this if I had known Electron adds a soft cap on memory. I read a few Github issues and it doesn't seem to be a definite number for the soft cap, however, I found this little flag snippet that will allow a user to increase the memory electron uses. `--js-flags="--max-old-space-size=4096"` 
I understand that by saying no sleep we are explicitly stating we don't ever want the application to sleep, however, do you think there would still be a way to clean the ram usage without sleeping the application?

**Comment by Thomas101** [comment]
We've managed to replicate what we think is the same issue here now. We've got a big memory box and after two days of running it appears to be showing something similar, so we'll do some more testing on that!

I'm also looking at putting a call in to manually GC - although when we tested it only seemed to make a negligible difference

**Comment by Thomas101** [comment]
Hi, we've released a fix for the calendar notifications in 3.5.2 which looks to be the cause of this. It would be good to know if you're still seeing the issue. If so there's some extra debugging tools we added too so we can give a few more things a try. Thanks

**Comment by Thomas101** [comment]
Hi, this has just been released in the latest version of Wavebox. You can get the update by using the Application/File menu and clicking `Check for Update` or by downloading the latest version from our [download page](https://wavebox.io/download). Thanks!
